lib/prawn/svg/gradient_renderer.rb

```diff
@@ -0,0 +1,310 @@
+class Prawn::SVG::GradientRenderer
+  include Prawn::SVG::TransformUtils
+
+  def initialize(prawn, draw_type, from:, to:, stops:, matrix: nil, r1: nil, r2: nil, wrap: :pad, bounding_box: nil)
+    @prawn = prawn
+    @draw_type = draw_type
+    @from = from
+    @to = to
+    @bounding_box = bounding_box
+
+    if r1
+      @shading_type = 3
+      @coordinates = [*from, r1, *to, r2]
+    else
+      @shading_type = 2
+      @coordinates = [*from, *to]
+    end
+
+    @stop_offsets, @color_stops, @opacity_stops = process_stop_arguments(stops)
+    @gradient_matrix = matrix ? load_matrix(matrix) : Matrix.identity(3)
+    @wrap = wrap
+  end
+
+  def draw
+    # If we need transparency, add an ExtGState to the page and enable it.
+    if opacity_stops
+      prawn.page.ext_gstates["PSVG-ExtGState-#{key}"] = create_transparency_graphics_state
+      prawn.renderer.add_content("/PSVG-ExtGState-#{key} gs")
+    end
+
+    # Add pattern to the PDF page resources dictionary.
+    prawn.page.resources[:Pattern] ||= {}
+    prawn.page.resources[:Pattern]["PSVG-Pattern-#{key}"] = create_gradient_pattern
+
+    # Finally set the pattern with the drawing operator for fill/stroke.
+    prawn.send(:set_color_space, draw_type, :Pattern)
+    draw_operator = draw_type == :fill ? 'scn' : 'SCN'
+    prawn.renderer.add_content("/PSVG-Pattern-#{key} #{draw_operator}")
+  end
+
+  private
+
+  attr_reader :prawn, :draw_type, :shading_type, :coordinates, :from, :to,
+    :stop_offsets, :color_stops, :opacity_stops, :gradient_matrix, :wrap, :bounding_box
+
+  def key
+    @key ||= Digest::SHA1.hexdigest([
+      draw_type, shading_type, coordinates, stop_offsets, color_stops, opacity_stops, gradient_matrix
```

Comment:
It looks like `wrap` is also used in the `create_gradient_pattern` method which is then cached under a key, should it be included in this list of digested variables?

Reply:
Good catch. You are right, both `wrap` and the (optional) `bounding_box` should be included to correctly hash the gradient. I have now added those.